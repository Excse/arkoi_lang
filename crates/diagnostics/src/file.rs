#[cfg(feature = "serialize")]
use serde::Serialize;

use crate::positional::Span;

pub type FileID = u32;

#[cfg_attr(feature = "serialize", derive(Serialize))]
#[derive(Debug)]
pub struct File {
    pub(crate) path: String,
    pub(crate) source: String,
    pub(crate) lines: Vec<Span>,
}

impl File {
    fn new<S>(path: S, source: S) -> Self
    where
        S: Into<String>,
    {
        let source = source.into();
        let path = path.into();

        let lines = File::line_ranges(&source);

        File {
            path,
            source,
            lines,
        }
    }

    fn line_ranges(source: &String) -> Vec<Span> {
        let mut lines = Vec::new();

        let mut start = 0;
        for (index, test) in source.match_indices('\n') {
            let span = Span::new(start, index);
            lines.push(span);
            start = index + 1;
        }

        if start != source.len() {
            let span = Span::new(start, source.len());
            lines.push(span)
        }

        lines
    }

    pub fn find_line_span(&self, char_span: &Span) -> Option<Span> {
        let start = self
            .lines
            .iter()
            .position(|span| span.is_inside(char_span.start))?;
        let end = self
            .lines
            .iter()
            .position(|span| span.is_inside(char_span.end))?;

        Some(Span::new(start, end))
    }

    pub fn slice(&self, span: &Span) -> Option<&str> {
        if span.end > self.source.len() {
            return None;
        }

        Some(&self.source[span.start..span.end])
    }
}

#[cfg_attr(feature = "serialize", derive(Serialize))]
#[derive(Debug)]
pub struct Files {
    files: Vec<File>,
}

impl Default for Files {
    fn default() -> Self {
        Files::new()
    }
}

impl Files {
    pub fn new() -> Self {
        Files { files: Vec::new() }
    }

    pub fn add<S>(&mut self, path: S, source: S) -> FileID
    where
        S: Into<String>,
    {
        self.files.push(File::new(path, source));

        u32::try_from(self.files.len()).expect("Shouldn't exceed the u32 bounds.")
    }

    pub fn get(&self, file_id: FileID) -> Option<&File> {
        self.files.get((file_id - 1) as usize)
    }

    pub fn path(&self, file_id: FileID) -> Option<&str> {
        let file = self.get(file_id)?;
        Some(&file.path)
    }

    pub fn source(&self, file_id: FileID) -> Option<&str> {
        let file = self.get(file_id)?;
        Some(&file.source)
    }

    pub fn slice(&self, file_id: FileID, span: &Span) -> Option<&str> {
        let file = self.get(file_id)?;
        file.slice(span)
    }

    pub fn find_line_span(&self, file_id: FileID, char_span: &Span) -> Option<Span> {
        let file = self.get(file_id)?;
        file.find_line_span(char_span)
    }
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn files() {
        let mut files = Files::default();

        let file_id = files.add("test_1.ark", "Hello World from 1!");
        assert_eq!(file_id, 1);
        assert_eq!(files.path(file_id), Some("test_1.ark"));
        assert_eq!(files.source(file_id), Some("Hello World from 1!"));
        let span = Span::new(17, 18);
        assert_eq!(files.slice(file_id, &span), Some("1"));

        let file_id = files.add("test_2.ark", "Hello World from 2!");
        assert_eq!(file_id, 2);
        assert_eq!(files.path(file_id), Some("test_2.ark"));
        assert_eq!(files.source(file_id), Some("Hello World from 2!"));
        let span = Span::new(17, 18);
        assert_eq!(files.slice(file_id, &span), Some("2"));
    }

    #[test]
    fn lines() {
        let mut files = Files::new();

        let file_id = files.add("test.ark", "Hello\nWorld\n!");
        let file = files.get(file_id).unwrap();

        assert_eq!(
            file.lines,
            vec![Span::new(0, 5), Span::new(6, 11), Span::new(12, 13)]
        );

        let lines = file
            .lines
            .iter()
            .map(|span| file.slice(span).unwrap())
            .collect::<Vec<&str>>();
        assert_eq!(lines, vec!["Hello", "World", "!"]);
    }

    #[test]
    fn line_spans() {
        let mut files = Files::new();

        let file_id = files.add(
            "test.ark",
            "Hello World!\nWhat is\nup?\nHave a great day :)",
        );
        let file = files.get(file_id).unwrap();

        let span = Span::new(0, 4);
        assert_eq!(file.find_line_span(&span), Some(Span::new(0, 0)));
        assert_eq!(file.slice(&span), Some("Hell"));

        let span = Span::new(0, 13);
        assert_eq!(file.find_line_span(&span), Some(Span::new(0, 1)));
        assert_eq!(file.slice(&span), Some("Hello World!\n"));

        let span = Span::new(13, 20);
        assert_eq!(file.find_line_span(&span), Some(Span::new(1, 1)));
        assert_eq!(file.slice(&span), Some("What is"));

        let span = Span::new(19, 23);
        assert_eq!(file.find_line_span(&span), Some(Span::new(1, 2)));
        assert_eq!(file.slice(&span), Some("s\nup"));
    }
}
