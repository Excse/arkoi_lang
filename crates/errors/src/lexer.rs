use diagnostics::{
    file::{FileID, Files},
    positional::Spannable,
    report::{LabelBuilder, Report, ReportBuilder, Serverity},
};

pub fn didnt_expect(
    files: &Files,
    file_id: FileID,
    got: Spannable<char>,
    expected: String,
) -> Report {
    let report_message = format!(
        "Expected to find '[{}]' but instead got '[{}]'.",
        expected, got.content,
    );
    let label_message = format!("Expected '[{}]' instead of this character.", expected);

    ReportBuilder::default()
        .message(report_message)
        .code(1)
        .serverity(Serverity::Error)
        .label(
            LabelBuilder::default()
                .message(label_message)
                .file(file_id)
                .span(got.span)
                .build(files)
                .unwrap(),
        )
        .build()
        .unwrap()
}
