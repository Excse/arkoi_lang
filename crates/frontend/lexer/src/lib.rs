mod cursor;
pub mod error;
pub mod iterator;
pub mod lexer;
pub mod token;

pub use crate::lexer::*;
