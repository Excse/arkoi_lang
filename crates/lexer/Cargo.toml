[package]
name = "lexer"
authors = { workspace = true }
categories = { workspace = true }
edition = { workspace = true }
version = { workspace = true }

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
diagnostics = { path = "../diagnostics" }

insta = { version = "1.31.0", features = ["yaml"] }
lasso = { version = "0.7.2", features = ["serialize"] }
serde = { version = "1.0.188", features = ["derive"], optional = true }

[features]
default = ["serialize"]
serialize = ["dep:serde"]
