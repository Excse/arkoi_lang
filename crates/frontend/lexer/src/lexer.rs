#[cfg(feature = "serialize")]
use serde::Serialize;

use lasso::Rodeo;

use crate::cursor::Cursor;
use crate::error::{InternalError, LexerError, Result};
use crate::token::TokenKind;
use diagnostics::file::{FileID, Files};

#[cfg_attr(feature = "serialize", derive(Serialize))]
#[derive(Debug)]
pub struct Lexer<'a> {
    pub(crate) cursor: Cursor<'a>,
    pub(crate) interner: &'a mut Rodeo,
    pub(crate) file_id: FileID,
    pub errors: Vec<LexerError>,
}

impl<'a> Lexer<'a> {
    pub fn new(files: &'a Files, file_id: FileID, interner: &'a mut Rodeo) -> Lexer<'a> {
        Lexer {
            cursor: Cursor::new(file_id, files),
            interner,
            errors: Vec::new(),
            file_id,
        }
    }

    pub(crate) fn next_token_kind(&mut self) -> Result<TokenKind> {
        let current = match self.cursor.peek() {
            Some(char) => char,
            None => return Err(LexerError::Internal(InternalError::UnexpectedEOF)),
        };

        self.cursor.mark_start();
        match current {
            char if char.is_alphabetic() => self.read_identifier(),
            char if char.is_numeric() => self.read_number(),
            '"' => self.read_string(),
            _ => self.read_symbol(),
        }
    }

    fn read_symbol(&mut self) -> Result<TokenKind> {
        let mut token = match self.cursor.try_consume() {
            Some(char) if char.is_whitespace() => self.next_token_kind()?,
            Some('{') => TokenKind::Brace(true),
            Some('}') => TokenKind::Brace(false),
            Some('(') => TokenKind::Parent(true),
            Some(')') => TokenKind::Parent(false),
            Some('[') => TokenKind::Bracket(true),
            Some(']') => TokenKind::Bracket(false),
            Some('@') => TokenKind::At,
            Some(',') => TokenKind::Comma,
            Some('.') => TokenKind::Period,
            Some('+') => TokenKind::Plus,
            Some('-') => TokenKind::Minus,
            Some('*') => TokenKind::Asterisk,
            Some('/') => TokenKind::Slash,
            Some('<') => TokenKind::Less,
            Some('>') => TokenKind::Greater,
            Some('=') => TokenKind::Eq,
            Some('!') => TokenKind::Apostrophe,
            Some(';') => TokenKind::Semicolon,
            Some(char) => TokenKind::Unknown(char),
            None => return Err(LexerError::Internal(InternalError::UnexpectedEOF)),
        };

        let current = match self.cursor.peek() {
            Some(char) => char,
            None => return Ok(token),
        };

        token = match (token, current) {
            (TokenKind::Plus, '=') => TokenKind::PlusEq,
            (TokenKind::Minus, '=') => TokenKind::MinusEq,
            (TokenKind::Asterisk, '=') => TokenKind::AsteriskEq,
            (TokenKind::Slash, '=') => TokenKind::SlashEq,
            (TokenKind::Less, '=') => TokenKind::LessEq,
            (TokenKind::Greater, '=') => TokenKind::GreaterEq,
            (TokenKind::Eq, '=') => TokenKind::EqEq,
            (TokenKind::Apostrophe, '=') => TokenKind::NotEq,
            (token, _) => return Ok(token),
        };

        self.cursor.try_consume();

        Ok(token)
    }

    fn read_identifier(&mut self) -> Result<TokenKind> {
        self.cursor.eat_if(char::is_alphabetic, "a-zA-Z")?;

        self.cursor
            .eat_while(|char| char.is_alphanumeric() || char == '_');

        Ok(match self.cursor.as_str() {
            "true" => TokenKind::True,
            "false" => TokenKind::False,
            "struct" => TokenKind::Struct,
            "return" => TokenKind::Return,
            "let" => TokenKind::Let,
            "self" => TokenKind::Self_,
            "fun" => TokenKind::Fun,
            "u8" => TokenKind::U8,
            "i8" => TokenKind::I8,
            "u16" => TokenKind::U16,
            "i16" => TokenKind::I16,
            "u32" => TokenKind::U32,
            "i32" => TokenKind::I32,
            "u64" => TokenKind::U64,
            "i64" => TokenKind::I64,
            "usize" => TokenKind::USize,
            "isize" => TokenKind::ISize,
            "f32" => TokenKind::F32,
            "f64" => TokenKind::F64,
            "bool" => TokenKind::Bool,
            _ => TokenKind::Id,
        })
    }

    fn read_number(&mut self) -> Result<TokenKind> {
        self.cursor.eat_if(char::is_numeric, "0-9")?;

        self.cursor.eat_while(char::is_numeric);

        if self.cursor.try_eat('.').is_ok() {
            self.cursor.eat_while(char::is_numeric);
            Ok(TokenKind::Decimal)
        } else {
            Ok(TokenKind::Int)
        }
    }

    fn read_string(&mut self) -> Result<TokenKind> {
        self.cursor.try_eat('"')?;

        self.cursor
            .eat_windowed_while(|prev, curr| curr != '"' || prev == '\\');

        self.cursor.try_eat('"')?;

        Ok(TokenKind::String)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::token::Token;

    macro_rules! test_token {
        (FAIL: $name:ident, $func:ident, $source:expr) => {
            #[test]
            fn $name() {
                let mut files = Files::default();
                let file_id = files.add("test.ark", $source);

                let mut interner = Rodeo::default();

                let mut lexer = Lexer::new(&files, file_id, &mut interner);
                let token = lexer.$func();
                assert!(token.is_err(), "{:?} should be an error", token);
            }
        };
        ($name:ident, $source:expr => $expected:expr) => {
            #[test]
            fn $name() {
                let mut files = Files::default();
                let file_id = files.add("test.ark", $source);

                let mut interner = Rodeo::default();

                let mut lexer = Lexer::new(&files, file_id, &mut interner);
                let expected = TokenKind::from($expected);
                let token = lexer.next_token_kind().unwrap();
                assert!(token == expected, "Input was {:?}", $source);
            }
        };
    }

    test_token!(success_decimal, "4.2" => TokenKind::Decimal);
    test_token!(success_integer, "42" => TokenKind::Int);
    test_token!(FAIL: fail_number, read_number, "number");

    test_token!(success_string, "\"Hello World!\"" => TokenKind::String);
    test_token!(FAIL: fail_string, read_string, "Hello World!");

    test_token!(success_true, "true" => TokenKind::True);
    test_token!(success_false, "false" => TokenKind::False);

    test_token!(success_obracket, "{" => TokenKind::Brace(true));
    test_token!(success_cbracket, "}" => TokenKind::Brace(false));
    test_token!(success_oparent, "(" => TokenKind::Parent(true));
    test_token!(success_cparent, ")" => TokenKind::Parent(false));
    test_token!(success_at, "@" => TokenKind::At);
    test_token!(success_apostrophe, "!" => TokenKind::Apostrophe);
    test_token!(success_comma, "," => TokenKind::Comma);
    test_token!(success_period, "." => TokenKind::Period);
    test_token!(success_semicolon, ";" => TokenKind::Semicolon);
    test_token!(success_addassign, "+=" => TokenKind::PlusEq);
    test_token!(success_plus, "+" => TokenKind::Plus);
    test_token!(success_minusassing, "-=" => TokenKind::MinusEq);
    test_token!(success_minus, "-" => TokenKind::Minus);
    test_token!(success_mulassign, "*=" => TokenKind::AsteriskEq);
    test_token!(success_asterisk, "*" => TokenKind::Asterisk);
    test_token!(success_divassign, "/=" => TokenKind::SlashEq);
    test_token!(success_slash, "/" => TokenKind::Slash);
    test_token!(success_lessequal, "<=" => TokenKind::LessEq);
    test_token!(success_less, "<" => TokenKind::Less);
    test_token!(success_greaterequal, ">=" => TokenKind::GreaterEq);
    test_token!(success_greater, ">" => TokenKind::Greater);
    test_token!(success_equal, "==" => TokenKind::EqEq);
    test_token!(success_notequal, "!=" => TokenKind::NotEq);
    test_token!(success_assign, "=" => TokenKind::Eq);

    test_token!(success_self, "self" => TokenKind::Self_);
    test_token!(success_u8, "u8" => TokenKind::U8);
    test_token!(success_i8, "i8" => TokenKind::I8);
    test_token!(success_u16, "u16" => TokenKind::U16);
    test_token!(success_i16, "i16" => TokenKind::I16);
    test_token!(success_u32, "u32" => TokenKind::U32);
    test_token!(success_i32, "i32" => TokenKind::I32);
    test_token!(success_u64, "u64" => TokenKind::U64);
    test_token!(success_i64, "i64" => TokenKind::I64);
    test_token!(success_usize, "usize" => TokenKind::USize);
    test_token!(success_isize, "isize" => TokenKind::ISize);
    test_token!(success_f32, "f32" => TokenKind::F32);
    test_token!(success_f64, "f64" => TokenKind::F64);
    test_token!(success_bool, "bool" => TokenKind::Bool);

    macro_rules! insta_test {
        ($name:ident, $path:expr) => {
            #[derive(Serialize)]
            struct InstaSnapshot<'a> {
                tokens: &'a [Token],
                interner: &'a Rodeo,
            }

            #[test]
            fn $name() {
                let mut files = Files::default();

                let source = std::fs::read_to_string($path).expect("Couldn't read the file.");
                let file_id = files.add($path, &source);

                let mut interner = Rodeo::default();

                let lexer = Lexer::new(&files, file_id, &mut interner);
                let iterator = lexer.into_iter();
                let tokens = iterator.collect::<Vec<Token>>();

                insta::assert_yaml_snapshot!(InstaSnapshot {
                    tokens: &tokens,
                    interner: &interner,
                });
            }
        };
    }

    insta_test!(insta_test, "test_files/insta_test.ark");
}
