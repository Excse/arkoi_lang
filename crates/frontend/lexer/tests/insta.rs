#[cfg(feature = "serialize")]
use serde::Serialize;

use std::cell::RefCell;
use std::rc::Rc;

use lasso::Rodeo;

use diagnostics::file::Files;
use lexer::{token::Token, Lexer};

macro_rules! insta_test {
    ($name:ident, $path:expr) => {
        #[derive(Serialize)]
        struct InstaSnapshot<'a> {
            tokens: &'a [Token],
            interner: &'a Rodeo,
        }

        #[test]
        fn $name() {
            let mut files = Files::default();

            let source = std::fs::read_to_string($path).expect("Couldn't read the file.");
            let file_id = files.add($path, &source);

            let interner = Rc::new(RefCell::new(Rodeo::default()));

            let lexer = Lexer::new(&files, file_id, interner.clone());
            let iterator = lexer.into_iter();
            let tokens = iterator.collect::<Vec<Token>>();

            insta::assert_yaml_snapshot!(InstaSnapshot {
                tokens: &tokens,
                interner: &interner.borrow(),
            });
        }
    };
}

insta_test!(insta_test, "test_files/insta_test.ark");
