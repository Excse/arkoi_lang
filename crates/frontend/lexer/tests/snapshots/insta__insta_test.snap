---
source: crates/frontend/lexer/tests/insta.rs
expression: "InstaSnapshot { tokens: &tokens, interner: &interner.borrow() }"
---
tokens:
  - file_id: 1
    value: ~
    kind: struct
    span:
      span:
        start: 0
        end: 6
      file_id: 1
  - file_id: 1
    value:
      String: 1
    kind: identifier
    span:
      span:
        start: 7
        end: 10
      file_id: 1
  - file_id: 1
    value: ~
    kind: "{"
    span:
      span:
        start: 11
        end: 12
      file_id: 1
  - file_id: 1
    value:
      String: 2
    kind: identifier
    span:
      span:
        start: 14
        end: 17
      file_id: 1
  - file_id: 1
    value: ~
    kind: "@"
    span:
      span:
        start: 18
        end: 19
      file_id: 1
  - file_id: 1
    value: ~
    kind: i32
    span:
      span:
        start: 19
        end: 22
      file_id: 1
  - file_id: 1
    value: ~
    kind: "!"
    span:
      span:
        start: 22
        end: 23
      file_id: 1
  - file_id: 1
    value: ~
    kind: ","
    span:
      span:
        start: 23
        end: 24
      file_id: 1
  - file_id: 1
    value:
      String: 3
    kind: identifier
    span:
      span:
        start: 26
        end: 29
      file_id: 1
  - file_id: 1
    value: ~
    kind: "@"
    span:
      span:
        start: 30
        end: 31
      file_id: 1
  - file_id: 1
    value: ~
    kind: i64
    span:
      span:
        start: 31
        end: 34
      file_id: 1
  - file_id: 1
    value: ~
    kind: ","
    span:
      span:
        start: 34
        end: 35
      file_id: 1
  - file_id: 1
    value: ~
    kind: "}"
    span:
      span:
        start: 36
        end: 37
      file_id: 1
  - file_id: 1
    value: ~
    kind: fun
    span:
      span:
        start: 39
        end: 42
      file_id: 1
  - file_id: 1
    value:
      String: 4
    kind: identifier
    span:
      span:
        start: 43
        end: 47
      file_id: 1
  - file_id: 1
    value: ~
    kind: (
    span:
      span:
        start: 47
        end: 48
      file_id: 1
  - file_id: 1
    value: ~
    kind: self
    span:
      span:
        start: 48
        end: 52
      file_id: 1
  - file_id: 1
    value: ~
    kind: "@"
    span:
      span:
        start: 53
        end: 54
      file_id: 1
  - file_id: 1
    value:
      String: 1
    kind: identifier
    span:
      span:
        start: 54
        end: 57
      file_id: 1
  - file_id: 1
    value: ~
    kind: "!"
    span:
      span:
        start: 57
        end: 58
      file_id: 1
  - file_id: 1
    value: ~
    kind: )
    span:
      span:
        start: 58
        end: 59
      file_id: 1
  - file_id: 1
    value: ~
    kind: "@"
    span:
      span:
        start: 60
        end: 61
      file_id: 1
  - file_id: 1
    value:
      String: 5
    kind: identifier
    span:
      span:
        start: 61
        end: 65
      file_id: 1
  - file_id: 1
    value: ~
    kind: "{"
    span:
      span:
        start: 66
        end: 67
      file_id: 1
  - file_id: 1
    value: ~
    kind: self
    span:
      span:
        start: 69
        end: 73
      file_id: 1
  - file_id: 1
    value: ~
    kind: "."
    span:
      span:
        start: 73
        end: 74
      file_id: 1
  - file_id: 1
    value:
      String: 2
    kind: identifier
    span:
      span:
        start: 74
        end: 77
      file_id: 1
  - file_id: 1
    value: ~
    kind: "="
    span:
      span:
        start: 78
        end: 79
      file_id: 1
  - file_id: 1
    value:
      Integer: 42
    kind: int
    span:
      span:
        start: 80
        end: 82
      file_id: 1
  - file_id: 1
    value: ~
    kind: ;
    span:
      span:
        start: 82
        end: 83
      file_id: 1
  - file_id: 1
    value: ~
    kind: "}"
    span:
      span:
        start: 84
        end: 85
      file_id: 1
  - file_id: 1
    value: ~
    kind: struct
    span:
      span:
        start: 87
        end: 93
      file_id: 1
  - file_id: 1
    value:
      String: 6
    kind: identifier
    span:
      span:
        start: 94
        end: 98
      file_id: 1
  - file_id: 1
    value: ~
    kind: "{"
    span:
      span:
        start: 99
        end: 100
      file_id: 1
  - file_id: 1
    value:
      String: 7
    kind: identifier
    span:
      span:
        start: 103
        end: 108
      file_id: 1
  - file_id: 1
    value: ~
    kind: "@"
    span:
      span:
        start: 109
        end: 110
      file_id: 1
  - file_id: 1
    value: ~
    kind: i64
    span:
      span:
        start: 110
        end: 113
      file_id: 1
  - file_id: 1
    value: ~
    kind: "!"
    span:
      span:
        start: 113
        end: 114
      file_id: 1
  - file_id: 1
    value: ~
    kind: ","
    span:
      span:
        start: 114
        end: 115
      file_id: 1
  - file_id: 1
    value:
      String: 8
    kind: identifier
    span:
      span:
        start: 117
        end: 122
      file_id: 1
  - file_id: 1
    value: ~
    kind: "@"
    span:
      span:
        start: 123
        end: 124
      file_id: 1
  - file_id: 1
    value: ~
    kind: i64
    span:
      span:
        start: 124
        end: 127
      file_id: 1
  - file_id: 1
    value: ~
    kind: "!"
    span:
      span:
        start: 127
        end: 128
      file_id: 1
  - file_id: 1
    value: ~
    kind: ","
    span:
      span:
        start: 128
        end: 129
      file_id: 1
  - file_id: 1
    value:
      String: 2
    kind: identifier
    span:
      span:
        start: 131
        end: 134
      file_id: 1
  - file_id: 1
    value: ~
    kind: "@"
    span:
      span:
        start: 135
        end: 136
      file_id: 1
  - file_id: 1
    value:
      String: 1
    kind: identifier
    span:
      span:
        start: 136
        end: 139
      file_id: 1
  - file_id: 1
    value: ~
    kind: ","
    span:
      span:
        start: 139
        end: 140
      file_id: 1
  - file_id: 1
    value: ~
    kind: "}"
    span:
      span:
        start: 141
        end: 142
      file_id: 1
  - file_id: 1
    value: ~
    kind: fun
    span:
      span:
        start: 144
        end: 147
      file_id: 1
  - file_id: 1
    value:
      String: 4
    kind: identifier
    span:
      span:
        start: 148
        end: 152
      file_id: 1
  - file_id: 1
    value: ~
    kind: (
    span:
      span:
        start: 152
        end: 153
      file_id: 1
  - file_id: 1
    value: ~
    kind: self
    span:
      span:
        start: 153
        end: 157
      file_id: 1
  - file_id: 1
    value: ~
    kind: "@"
    span:
      span:
        start: 158
        end: 159
      file_id: 1
  - file_id: 1
    value:
      String: 6
    kind: identifier
    span:
      span:
        start: 159
        end: 163
      file_id: 1
  - file_id: 1
    value: ~
    kind: "!"
    span:
      span:
        start: 163
        end: 164
      file_id: 1
  - file_id: 1
    value: ~
    kind: )
    span:
      span:
        start: 164
        end: 165
      file_id: 1
  - file_id: 1
    value: ~
    kind: "@"
    span:
      span:
        start: 166
        end: 167
      file_id: 1
  - file_id: 1
    value:
      String: 5
    kind: identifier
    span:
      span:
        start: 167
        end: 171
      file_id: 1
  - file_id: 1
    value: ~
    kind: "{"
    span:
      span:
        start: 172
        end: 173
      file_id: 1
  - file_id: 1
    value: ~
    kind: self
    span:
      span:
        start: 175
        end: 179
      file_id: 1
  - file_id: 1
    value: ~
    kind: "."
    span:
      span:
        start: 179
        end: 180
      file_id: 1
  - file_id: 1
    value:
      String: 7
    kind: identifier
    span:
      span:
        start: 180
        end: 185
      file_id: 1
  - file_id: 1
    value: ~
    kind: "="
    span:
      span:
        start: 186
        end: 187
      file_id: 1
  - file_id: 1
    value:
      Integer: 21
    kind: int
    span:
      span:
        start: 188
        end: 190
      file_id: 1
  - file_id: 1
    value: ~
    kind: ;
    span:
      span:
        start: 190
        end: 191
      file_id: 1
  - file_id: 1
    value: ~
    kind: self
    span:
      span:
        start: 193
        end: 197
      file_id: 1
  - file_id: 1
    value: ~
    kind: "."
    span:
      span:
        start: 197
        end: 198
      file_id: 1
  - file_id: 1
    value:
      String: 8
    kind: identifier
    span:
      span:
        start: 198
        end: 203
      file_id: 1
  - file_id: 1
    value: ~
    kind: "="
    span:
      span:
        start: 204
        end: 205
      file_id: 1
  - file_id: 1
    value:
      Integer: 21
    kind: int
    span:
      span:
        start: 206
        end: 208
      file_id: 1
  - file_id: 1
    value: ~
    kind: ;
    span:
      span:
        start: 208
        end: 209
      file_id: 1
  - file_id: 1
    value: ~
    kind: "}"
    span:
      span:
        start: 210
        end: 211
      file_id: 1
  - file_id: 1
    value: ~
    kind: fun
    span:
      span:
        start: 213
        end: 216
      file_id: 1
  - file_id: 1
    value:
      String: 9
    kind: identifier
    span:
      span:
        start: 217
        end: 220
      file_id: 1
  - file_id: 1
    value: ~
    kind: (
    span:
      span:
        start: 220
        end: 221
      file_id: 1
  - file_id: 1
    value: ~
    kind: self
    span:
      span:
        start: 221
        end: 225
      file_id: 1
  - file_id: 1
    value: ~
    kind: "@"
    span:
      span:
        start: 226
        end: 227
      file_id: 1
  - file_id: 1
    value:
      String: 6
    kind: identifier
    span:
      span:
        start: 227
        end: 231
      file_id: 1
  - file_id: 1
    value: ~
    kind: )
    span:
      span:
        start: 231
        end: 232
      file_id: 1
  - file_id: 1
    value: ~
    kind: "@"
    span:
      span:
        start: 233
        end: 234
      file_id: 1
  - file_id: 1
    value: ~
    kind: i64
    span:
      span:
        start: 234
        end: 237
      file_id: 1
  - file_id: 1
    value: ~
    kind: "{"
    span:
      span:
        start: 238
        end: 239
      file_id: 1
  - file_id: 1
    value: ~
    kind: self
    span:
      span:
        start: 241
        end: 245
      file_id: 1
  - file_id: 1
    value: ~
    kind: "."
    span:
      span:
        start: 245
        end: 246
      file_id: 1
  - file_id: 1
    value:
      String: 7
    kind: identifier
    span:
      span:
        start: 246
        end: 251
      file_id: 1
  - file_id: 1
    value: ~
    kind: +
    span:
      span:
        start: 252
        end: 253
      file_id: 1
  - file_id: 1
    value: ~
    kind: self
    span:
      span:
        start: 254
        end: 258
      file_id: 1
  - file_id: 1
    value: ~
    kind: "."
    span:
      span:
        start: 258
        end: 259
      file_id: 1
  - file_id: 1
    value:
      String: 8
    kind: identifier
    span:
      span:
        start: 259
        end: 264
      file_id: 1
  - file_id: 1
    value: ~
    kind: +
    span:
      span:
        start: 265
        end: 266
      file_id: 1
  - file_id: 1
    value: ~
    kind: self
    span:
      span:
        start: 267
        end: 271
      file_id: 1
  - file_id: 1
    value: ~
    kind: "."
    span:
      span:
        start: 271
        end: 272
      file_id: 1
  - file_id: 1
    value:
      String: 2
    kind: identifier
    span:
      span:
        start: 272
        end: 275
      file_id: 1
  - file_id: 1
    value: ~
    kind: "."
    span:
      span:
        start: 275
        end: 276
      file_id: 1
  - file_id: 1
    value:
      String: 2
    kind: identifier
    span:
      span:
        start: 276
        end: 279
      file_id: 1
  - file_id: 1
    value: ~
    kind: +
    span:
      span:
        start: 280
        end: 281
      file_id: 1
  - file_id: 1
    value: ~
    kind: self
    span:
      span:
        start: 282
        end: 286
      file_id: 1
  - file_id: 1
    value: ~
    kind: "."
    span:
      span:
        start: 286
        end: 287
      file_id: 1
  - file_id: 1
    value:
      String: 2
    kind: identifier
    span:
      span:
        start: 287
        end: 290
      file_id: 1
  - file_id: 1
    value: ~
    kind: "."
    span:
      span:
        start: 290
        end: 291
      file_id: 1
  - file_id: 1
    value:
      String: 3
    kind: identifier
    span:
      span:
        start: 291
        end: 294
      file_id: 1
  - file_id: 1
    value: ~
    kind: "}"
    span:
      span:
        start: 295
        end: 296
      file_id: 1
  - file_id: 1
    value: ~
    kind: fun
    span:
      span:
        start: 298
        end: 301
      file_id: 1
  - file_id: 1
    value:
      String: 10
    kind: identifier
    span:
      span:
        start: 302
        end: 306
      file_id: 1
  - file_id: 1
    value: ~
    kind: (
    span:
      span:
        start: 306
        end: 307
      file_id: 1
  - file_id: 1
    value: ~
    kind: )
    span:
      span:
        start: 307
        end: 308
      file_id: 1
  - file_id: 1
    value: ~
    kind: "@"
    span:
      span:
        start: 309
        end: 310
      file_id: 1
  - file_id: 1
    value: ~
    kind: i32
    span:
      span:
        start: 310
        end: 313
      file_id: 1
  - file_id: 1
    value: ~
    kind: "{"
    span:
      span:
        start: 314
        end: 315
      file_id: 1
  - file_id: 1
    value: ~
    kind: let
    span:
      span:
        start: 317
        end: 320
      file_id: 1
  - file_id: 1
    value:
      String: 11
    kind: identifier
    span:
      span:
        start: 321
        end: 325
      file_id: 1
  - file_id: 1
    value: ~
    kind: "@"
    span:
      span:
        start: 326
        end: 327
      file_id: 1
  - file_id: 1
    value: ~
    kind: i32
    span:
      span:
        start: 327
        end: 330
      file_id: 1
  - file_id: 1
    value: ~
    kind: "="
    span:
      span:
        start: 331
        end: 332
      file_id: 1
  - file_id: 1
    value:
      Integer: 0
    kind: int
    span:
      span:
        start: 333
        end: 334
      file_id: 1
  - file_id: 1
    value: ~
    kind: ;
    span:
      span:
        start: 334
        end: 335
      file_id: 1
  - file_id: 1
    value:
      String: 11
    kind: identifier
    span:
      span:
        start: 337
        end: 341
      file_id: 1
  - file_id: 1
    value: ~
    kind: "}"
    span:
      span:
        start: 342
        end: 343
      file_id: 1
interner:
  - Foo
  - foo
  - bar
  - init
  - void
  - Test
  - hello
  - world
  - add
  - main
  - test

